docs/content-flow.md

```diff
@@ -0,0 +1,388 @@
+# Content Flow
+
+A major goal of containerd is to create a system wherein content can be used for executing containers.
+In order to execute on that flow, containerd requires content and to manage it.
+
+This document describes how content flows into containerd, how it is managed, and where it exists
+at each stage in the process. We use an example of going from a known image
+[docker.io/library/redis:5.0.9](https://hub.docker.com/layers/redis/library/redis/5.0.9/images/sha256-4ff8940144391ecd5e1632d0c427d95f4a8d2bb4a72b7e3898733352350d9ab3?context=explore) to explore the
+flow of content.
+
+## Content Areas
+
+Content exists in several areas in the containerd lifecycle:
+
+* OCI registry, for example [hub.docker.com](https://hub.docker.com) or [quay.io](https://quay.io)
+* containerd content store, under containerd's local storage space, for example, on a standard Linux installation at `/var/lib/containerd/io.containerd.content.v1.content`
+* snapshots, under containerd's local storage space, for example, on a standard Linux installation at `/var/lib/containerd/io.containerd.snapshotter.v1.<type>`. For an overlayfs snapshotter, that would be at `/var/lib/containerd/io.containerd.snapshotter.v1.overlayfs`
+
+In order to create a container, the following must occur:
+
+1. The image and all its content must be loaded into the content store. This normally happens via download from the OCI registry, but you can load content in directly as well.
+1. Committed snapshots must be created from each layer of content for the image.
+1. An active snapshot must be created on top of the final layer of content for the image.
+
+A container now can be created, with its root filesystem at the active snapshot.
+
+The rest of this document looks at the content in each area in detail, and how they relate to one another.
+
+### Image Format
+
+Images in a registry normally are stored in the following format. An "image" is actually just a json document
```

Comment:
Can we remove "actually just"?

Maybe some wording changes:

```
An "image" is comprised of a json document, called a manifest, that describes the image including what platform it targets and pointers to other content hashes that make up the image.
```

Reply:
Also what may help here is describing the process of converting an image reference, such s `redis:5.0.9` to an image on disk.
I think this would give some continuity to the process... just as a rough example (maybe some of it is too much include, but at least it's a general flow):

We have an image reference `redis:5.0.9`, we ask the registry what does this refer to, it returns a descriptor containing a media type and a hash.
```json
{
    "mediaType": "application/vnd.docker.distribution.manifest.list.v2+json" 
    "digest": "SHA256:<some sha>",
    "size": <number of Bytes>
}
```

We then request the content referred to by that hash from the registry. We use the media type on the dscriptor to determine how to decode the content.
This one is an "index", which is a list of manifests, typically each manifest would refer to an image for a particular platform. The purpose of the index allows us to choose which manifest matches the platform we want.

... then on with picking the manifest, then the layers, etc and where they get stored in containerd and fetched in containerd.

Reply:
These are good suggestions. I have updated the document. Please take a look @cpuguy83